src/main/kotlin/app/CodeLongevity.kt

```diff
@@ -0,0 +1,231 @@
+
+    /**
+     * Id of the code line in a revision when the line was added. Used to
+     * update the line's lifetime computed in previous iterations.
+     * 
+     */
+    val oldId: String = ""
+
+    /**
+     * Id of the line in head revision (if this is an existing line) or
+     * a revision where the line was deleted.
+     */
+    val newId: String = ""
+
+    /**
+     * A line age in seconds.
+     */
+    val age = to.commit.getCommitTime() - from.commit.getCommitTime()
+
+    /**
+     * A pretty print of a code line; debugging.
+     */
+    fun printme() {
+        val fd = Date(from.commit.getCommitTime().toLong() * 1000).toLocaleString()
+        val fc = "${from.commit.getName()} '${from.commit.getShortMessage()}'"
+        val tc = "${to.commit.getName()} '${to.commit.getShortMessage()}'"
+        val td = Date(to.commit.getCommitTime().toLong() * 1000).toLocaleString()
+        println("Line '$text' - '${from.file}:${from.line}' added in $fc $fd")
+        println("  last known as '${to.file}:${to.line}' in $tc $td")
+    }
+}
+
+/**
+ * Used to compute age of code lines in the repo.
+ */
+class CodeLongevity(repoPath: String, tailRev: String) {
+    val repo = FileRepository(repoPath)
+    val head: RevCommit = RevWalk(repo).parseCommit(repo.resolve("HEAD"));
+    val tail: RevCommit? =
+        if (tailRev != "") RevWalk(repo).parseCommit(repo.resolve(tailRev))
+        else null
+
+    var codeLines: MutableList<CodeLine> = mutableListOf()
+
+    init {
+        compute()
+    }
+
+    fun test(email: String) {
+        var sum: Long = 0
+        var total: Long = 0;
+        for (line in codeLines) {
+            val author = line.from.commit.getAuthorIdent()
+            if (author.getEmailAddress() != email) {
+                continue
+            }
+            //println("got a line, age ${line.age}\n ${line.printme()}")
+            sum += line.age
+            total++
+        }
+
+        //println("All lines:")
+        //codeLines.forEach { line -> line.printme() }
+
+        var avg = if (total > 0) sum / total else 0
+        println("avg code line age for <$email> is ${avg} seconds")
+    }
+
+    /**
+     * Computes age of all lines in the repo through all revisions.
+     */
+    private fun compute() {
+        val treeWalk = TreeWalk(repo)
+        treeWalk.setRecursive(true)
+        treeWalk.addTree(head!!.getTree())
+
+        val files: MutableMap<String, ArrayList<RevCommitLine>> = mutableMapOf()
+
+        // Build a map of file names and their code lines.
+        while (treeWalk.next()) {
+            // TODO(alex): skip binary files.
+            val fileName = treeWalk.getPathString()
+            val fileLoader = repo.open(treeWalk.getObjectId(0))
+            val fileText = RawText(fileLoader.getBytes())
+            val linesNum = fileText.size()
+            var lines = ArrayList<RevCommitLine>(linesNum)
+            var idx = 0
+            while (idx < fileText.size()) {
```

Comment:
as far as I know - for (variable in collection), they probably have a fancy way to create an array for indices, not sure though if they unwrap into same effective code. I'll leave that on @anatolykislov's plate if you don't mind.